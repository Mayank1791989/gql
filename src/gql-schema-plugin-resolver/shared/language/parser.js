/* @flow strict */
import { Source } from 'graphql';
import { type GQLPosition } from 'gql-shared/types';
import { type ResolverDocumentNode } from './ast';
import { type ResolverToken } from './token';

export interface IResolverParser {
  parse(source: Source): ResolverDocumentNode;
  getTokenAtPosition(
    source: Source,
    position: GQLPosition,
  ): ResolverToken | null;
}
