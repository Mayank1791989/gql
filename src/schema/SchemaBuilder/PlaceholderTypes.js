/* @flow */
import {
  GQLScalarType,
  GQLObjectType,
  GQLInterfaceType,
} from '../../shared/GQLTypes';

const genUnknownScalarType = (name: string) =>
  new GQLScalarType(null, {
    name,
    description: `Unknown scalar type '${name}'`,
    serialize: () => null,
    parseValue: () => false,
    parseLiteral: () => false,
  });

// will be used to provide definition for unknown types
export const PLACEHOLDER_TYPES = {
  scalarType: genUnknownScalarType,
  inputType: genUnknownScalarType,
  outputType: genUnknownScalarType,
  objectType: (name: string) =>
    new GQLObjectType(null, {
      name,
      description: `unknown type '${name}'`,
      fields: {
        name: { type: genUnknownScalarType('unknown') },
      },
      interfaces: () => null,
    }),
  interfaceType: (name: string) =>
    new GQLInterfaceType(null, {
      name,
      description: `unknown interface type '${name}'`,
      fields: {
        unknown: { type: genUnknownScalarType('unknown') },
      },
      resolveType: () => {}, // eslint-disable-line no-empty-function
    }),
};
