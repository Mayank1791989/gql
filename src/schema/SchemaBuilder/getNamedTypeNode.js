/* @flow */
import {
  LIST_TYPE,
  NON_NULL_TYPE,
} from 'graphql/language/kinds';

import type {
  TypeNode,
  NamedTypeNode,
} from 'graphql/language/ast';

export default function getNamedTypeNode(typeNode: TypeNode): NamedTypeNode {
  let namedType = typeNode;
  while (namedType.kind === LIST_TYPE || namedType.kind === NON_NULL_TYPE) {
    namedType = namedType.type;
  }
  return namedType;
}
