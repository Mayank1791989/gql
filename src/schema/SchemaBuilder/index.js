/* @flow */
export { default } from './GQLSchemaBuilder';
export { default as SchemaBuilder } from './GQLSchemaBuilder';
