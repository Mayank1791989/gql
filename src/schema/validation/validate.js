/* @flow */
/* eslint-disable no-use-before-define */
import { visitInParallel, visit, GraphQLError } from 'graphql';
import type { DocumentNode } from 'graphql/language/ast';
import schemaRules from './rules';

import type { GQLSchema } from '../../shared/GQLTypes';

import { toGQLError, SEVERITY } from '../../shared/GQLError';
import type { GQLError } from '../../shared/GQLError';

export function validate(
  schema: GQLSchema,
  ast: DocumentNode,
): Array<GQLError> {
  const context = new ValidationContext(schema, ast);
  const visitors = schemaRules.map(rule => rule(context));
  visit(ast, visitInParallel(visitors));
  return context.getErrors();
}

// NOTE: keeping the structure of Validation context
// same as graphql-js ValidationContext so we can reuse
// graphql-js validation rule
class ValidationContext {
  _schema: GQLSchema;
  _ast: DocumentNode;
  _errors: Array<GQLError> = [];

  constructor(schema: GQLSchema, ast: DocumentNode) {
    this._schema = schema;
    this._ast = ast;
  }

  reportError(error: GraphQLError, severity: $Keys<typeof SEVERITY>) {
    this._errors.push(toGQLError(error, severity));
  }

  getErrors(): Array<GQLError> {
    return this._errors;
  }

  getSchema(): GQLSchema {
    return this._schema;
  }

  getDocument(): DocumentNode {
    return this._ast;
  }
}
