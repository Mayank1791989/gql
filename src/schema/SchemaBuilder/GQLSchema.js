/* @flow */
/**
 *  Note this is modified version of original GQLSchema
 *  1) Return Array<Errors> instead of throwing on first error.
 */

import {
  GraphQLObjectType,
  GraphQLNonNull,
} from 'graphql/type/definition';

import type { // eslint-disable-line no-duplicate-imports
  GraphQLAbstractType,
} from 'graphql/type/definition';

import type {
  TypeDefinitionNode,
  ASTNode,
} from 'graphql/language/ast';

import { specifiedDirectives } from 'graphql/type/directives';
import find from 'graphql/jsutils/find';
import invariant from 'graphql/jsutils/invariant';
import { isEqualType, isTypeSubTypeOf } from 'graphql/utilities/typeComparators';
import { newGQLError, SEVERITY } from '../../shared/GQLError';
import { PLACEHOLDER_TYPES } from './PlaceholderTypes';
import getNamedTypeNode from './getNamedTypeNode';

import type { GQLError } from '../../shared/GQLError';
import type { GQLSchema, GQLNamedType } from '../../shared/GQLTypes';
import { GQLObjectType, GQLInterfaceType, GQLUnionType, GQLDirective } from '../../shared/GQLTypes';

type GQLSchemaConfig = {
  query: ?GQLObjectType;
  mutation?: ?GQLObjectType;
  subscription?: ?GQLObjectType;
  types?: ?Array<GQLNamedType>;
  directives?: ?Array<GQLDirective>;
  nodeMap: { [name: string]: TypeDefinitionNode };
};

export default class _GQLSchema {
  _queryType: ?GQLObjectType;
  _mutationType: ?GQLObjectType;
  _subscriptionType: ?GQLObjectType;
  _directives: Array<GQLDirective>;
  _typeMap: { [typeName: string]: GQLNamedType } = {};
  _implementations: { [interfaceName: string]: Array<GQLObjectType> };
  _possibleTypeMap: ?{
    [abstractName: string]: { [possibleName: string]: boolean }
  };

  _nodeMap: { [name: string]: TypeDefinitionNode };
  _errors: Array<GQLError> = [];

  constructor(config: GQLSchemaConfig) {
    this._queryType = config.query;
    this._mutationType = config.mutation;
    this._subscriptionType = config.subscription;
    this._directives = config.directives || specifiedDirectives;
    this._implementations = {};
    this._nodeMap = config.nodeMap;

    // Build type map now to detect any errors within this schema.
    const types: Array<GQLNamedType> = config.types || [];

    // create typeMap and implementations map
    types.forEach((type) => {
      if (type) { this._typeMap[type.name] = type; }
      // // NOTE: dont remove type.getFields()
      // // calling to resolve thunks which will add more errors
      // // also wrapping in try catch to suppress errors thrown inside getFields function
      // // errors thrown by getFields is already handle while creating schema from AST
      if (type.getFields) {
        try {
          // $FlowDisableNextLine
          type.getFields(); // resolve thunk
        } catch (e) { } // eslint-disable-line no-empty
      }

      if (type instanceof GQLObjectType) {
        const _type = type;
        type.getInterfaces().forEach((iface) => {
          this._errors.push(
            // eslint-disable-next-line no-use-before-define
            ...assertObjectImplementsInterface((this: any), _type, iface),
          );
          const impls = this._implementations[iface.name];
          if (impls) {
            impls.push(_type);
          } else {
            this._implementations[iface.name] = [_type];
          }
        });
      }
    });
  }

  getQueryType(): ?GQLObjectType {
    return this._queryType;
  }

  getMutationType(): ?GQLObjectType {
    return this._mutationType;
  }

  getSubscriptionType(): ?GQLObjectType {
    return this._subscriptionType;
  }

  getTypeMap() {
    return this._typeMap;
  }

  getType(name: string): ?GQLNamedType {
    return this.getTypeMap()[name];
  }

  getTypeNode(name: string): ?TypeDefinitionNode {
    return this._nodeMap[name];
  }

  getTypeDependents(name: string): Array<ASTNode> {
    const type = this.getType(name);
    if (type) {
      return type.dependents || [];
    }
    return [];
  }

  getPossibleTypes(abstractType: any): Array<GQLObjectType> {
    if (abstractType instanceof GQLUnionType) {
      return abstractType.getTypes();
    }
    invariant(abstractType instanceof GQLInterfaceType);
    return (this._implementations[abstractType.name]: any);
  }

  isPossibleType(
    abstractType: GraphQLAbstractType,
    possibleType: GraphQLObjectType,
  ): boolean {
    let possibleTypeMap = this._possibleTypeMap;
    if (!possibleTypeMap) {
      this._possibleTypeMap = possibleTypeMap = Object.create(null);
    }

    if (!possibleTypeMap[abstractType.name]) {
      const possibleTypes = this.getPossibleTypes(abstractType);
      invariant(
        Array.isArray(possibleTypes),
        `Could not find possible implementing types for ${abstractType.name} ` +
        'in schema. Check that schema.types is defined and is an array of ' +
        'all possible types in the schema.',
      );
      possibleTypeMap[abstractType.name] =
        possibleTypes.reduce(
          (map, type) => ((map[type.name] = true), map), // eslint-disable-line
          Object.create(null),
        );
    }

    return Boolean(possibleTypeMap[abstractType.name][possibleType.name]);
  }

  getDirectives(): Array<GQLDirective> {
    return this._directives;
  }

  getDirective(name: string): ?GQLDirective {
    return find(this.getDirectives(), directive => directive.name === name);
  }
}

function assertObjectImplementsInterface(
  schema: GQLSchema,
  object: GQLObjectType,
  iface: GQLInterfaceType,
): Array<GQLError> {
  if (iface === PLACEHOLDER_TYPES.interfaceType) {
    return [];
  }

  const objectFieldMap = object.getFields();

  const ifaceFieldMap = iface.getFields();

  const missingFields = [];
  const errors = [];

  // Assert each interface field is implemented.
  Object.keys(ifaceFieldMap).forEach((fieldName) => {
    const objectField = objectFieldMap[fieldName];
    const ifaceField = ifaceFieldMap[fieldName];

    if (!objectField) {
      missingFields.push(ifaceField);
      return;
    }

    // Assert interface field type is satisfied by object field type, by being
    // a valid subtype. (covariant)
    if (!isTypeSubTypeOf((schema: any), (objectField.type: any), (ifaceField.type: any))) {
      errors.push(newGQLError(
        `${iface.name}.${fieldName} expects type "${String(ifaceField.type)}" ` +
        'but ' +
        `${object.name}.${fieldName} provides type "${String(objectField.type)}".`,
        [getNamedTypeNode(objectField.node.type)],
        SEVERITY.error,
      ));
    }

    // Assert each interface field arg is implemented.
    ifaceField.args.forEach((ifaceArg) => {
      const argName = ifaceArg.name;
      const objectArg = find(objectField.args, arg => arg.name === argName);

      // Assert interface field arg exists on object field.
      if (!objectArg) {
        errors.push(newGQLError(
          `${iface.name}.${fieldName} expects argument "${argName}" but ` +
          `${object.name}.${fieldName} does not provide it.`,
          [objectField.node],
          SEVERITY.error,
        ));
        return;
      }

      // Assert interface field arg type matches object field arg type.
      // (invariant)
      if (!isEqualType((ifaceArg.type: any), (objectArg.type: any))) {
        errors.push(newGQLError(
          `${iface.name}.${fieldName}(${argName}:) expects type ` +
          `"${String(ifaceArg.type)}" but ` +
          `${object.name}.${fieldName}(${argName}:) provides type ` +
          `"${String(objectArg.type)}".`,
          [getNamedTypeNode(objectArg.node.type)],
          SEVERITY.error,
        ));
      }
    });

    // Assert additional arguments must not be required.
    objectField.args.forEach((objectArg) => {
      const argName = objectArg.name;
      const ifaceArg = find(ifaceField.args, arg => arg.name === argName);
      if (!ifaceArg && objectArg.type instanceof GraphQLNonNull) {
        errors.push(newGQLError(
          `${object.name}.${fieldName}(${argName}:) is of required type ` +
          `"${String(objectArg.type)}" but is not also provided by the ` +
          `interface ${iface.name}.${fieldName}.`,
          [objectArg.node],
          SEVERITY.error,
        ));
      }
    });
  });

  if (missingFields.length > 0) {
    errors.push(newGQLError(
      `Missing interface fields [${missingFields.map(field => field.name).join(', ')}]`,
      [object.node],
      SEVERITY.error,
    ));
  }

  return errors;
}
