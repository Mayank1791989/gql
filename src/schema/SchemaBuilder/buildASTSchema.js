/* @flow */
/* eslint-disable no-use-before-define */
/**
 * NOTE: patched version
 */
import find from 'graphql/jsutils/find';
import invariant from 'graphql/jsutils/invariant';
import keyValMap from 'graphql/jsutils/keyValMap';
import { valueFromAST } from 'graphql/utilities/valueFromAST';
import { TokenKind } from 'graphql/language/lexer';
import { getArgumentValues } from 'graphql/execution/values';

import {
  LIST_TYPE,
  NON_NULL_TYPE,
  SCHEMA_DEFINITION,
  SCALAR_TYPE_DEFINITION,
  OBJECT_TYPE_DEFINITION,
  INTERFACE_TYPE_DEFINITION,
  ENUM_TYPE_DEFINITION,
  UNION_TYPE_DEFINITION,
  INPUT_OBJECT_TYPE_DEFINITION,
  DIRECTIVE_DEFINITION,
} from 'graphql/language/kinds';

import type {
  Location,
  ASTNode,
  DocumentNode,
  DirectiveNode,
  TypeNode,
  SchemaDefinitionNode,
  TypeDefinitionNode,
  ScalarTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  UnionTypeDefinitionNode,
  EnumTypeDefinitionNode,
  InputObjectTypeDefinitionNode,
  DirectiveDefinitionNode,
} from 'graphql/language/ast';

import GQLSchema from './GQLSchema';
import getNamedTypeNode from './getNamedTypeNode';

import {
  GQLString,
  GQLInt,
  GQLFloat,
  GQLBoolean,
  GQLID,
} from '../../shared/GQLTypes';

import {
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLNonNull,
  GraphQLList,
  isInputType,
  isOutputType,
} from 'graphql/type/definition';

import type { // eslint-disable-line no-duplicate-imports
  GraphQLType,
  GraphQLInputType,
  GraphQLOutputType,
} from 'graphql/type/definition';

import {
  GQLObjectType,
  GQLInputObjectType,
  GQLInterfaceType,
  GQLEnumType,
  GQLScalarType,
  GQLUnionType,
  GQLDirective,
  GQLSkipDirective,
  GQLIncludeDirective,
  GQLDeprecatedDirective,
} from '../../shared/GQLTypes';

import type { GQLNamedType, GQLType } from '../../shared/GQLTypes';

import type { // eslint-disable-line no-duplicate-imports
  DirectiveLocationEnum,
} from 'graphql/type/directives';

// import {
//   __Schema,
//   __Directive,
//   __DirectiveLocation,
//   __Type,
//   __Field,
//   __InputValue,
//   __EnumValue,
//   __TypeKind,
// } from 'graphql/type/introspection';

import { newGQLError, SEVERITY } from '../../shared/GQLError';
import type { GQLError } from '../../shared/GQLError';

import { PLACEHOLDER_TYPES } from './PlaceholderTypes';

function buildWrappedType(
  innerType: GraphQLType,
  inputTypeAST: TypeNode,
): GQLType {
  if (inputTypeAST.kind === LIST_TYPE) {
    return new GraphQLList((buildWrappedType(innerType, inputTypeAST.type): any));
  }
  if (inputTypeAST.kind === NON_NULL_TYPE) {
    const wrappedType = buildWrappedType(innerType, inputTypeAST.type);
    invariant(!(wrappedType instanceof GraphQLNonNull), 'No nesting nonnull.');
    return new GraphQLNonNull((wrappedType: any));
  }
  return (innerType: any);
}

export function buildASTSchema(ast: DocumentNode): { schema: GQLSchema, errors: Array<GQLError> } {
  let schemaDef: ?SchemaDefinitionNode;

  const errors: Array<GQLError> = [];
  const typeDefs: Array<TypeDefinitionNode> = [];
  const nodeMap: {[name: string]: TypeDefinitionNode} = Object.create(null);
  const nodeMapWithAllReferences: {[name: string]: Array<TypeDefinitionNode>} = Object.create(null);
  const directiveDefs: Array<DirectiveDefinitionNode> = [];

  for (let i = 0; i < ast.definitions.length; i += 1) {
    const d = ast.definitions[i];
    switch (d.kind) {
      case SCHEMA_DEFINITION:
        if (schemaDef) {
          errors.push(newGQLError(
            'Must provide only one schema definition.',
            [schemaDef, d],
            SEVERITY.error,
          ));
        }
        schemaDef = d;
        break;
      case SCALAR_TYPE_DEFINITION:
      case OBJECT_TYPE_DEFINITION:
      case INTERFACE_TYPE_DEFINITION:
      case ENUM_TYPE_DEFINITION:
      case UNION_TYPE_DEFINITION:
      case INPUT_OBJECT_TYPE_DEFINITION: {
        const name = d.name.value;
        if (!nodeMap[name]) {
          typeDefs.push(d);
          nodeMap[name] = d;
        }
        //  storing all reference to detect multiple defintition with same name
        if (!nodeMapWithAllReferences[name]) { nodeMapWithAllReferences[name] = []; }
        nodeMapWithAllReferences[name].push(d);
        break;
      }
      case DIRECTIVE_DEFINITION:
        directiveDefs.push(d);
        break;
      default:
    }
  }

  // error for multi same name typeDef
  Object.keys(nodeMapWithAllReferences).forEach((name) => {
    if (nodeMapWithAllReferences[name].length > 1) {
      errors.push(newGQLError(
        `Schema must contain unique named types but contains multiple types named "${name}".`,
        nodeMapWithAllReferences[name].map(typeDefAST => typeDefAST.name),
        SEVERITY.error,
      ));
    }
  });

  let queryTypeName;
  let mutationTypeName;
  let subscriptionTypeName;

  if (schemaDef) {
    schemaDef.operationTypes.forEach((operationType) => {
      const typeName = operationType.type.name.value;
      if (operationType.operation === 'query') {
        queryTypeName = typeName;
      } else if (operationType.operation === 'mutation') {
        mutationTypeName = typeName;
      } else if (operationType.operation === 'subscription') {
        subscriptionTypeName = typeName;
      }
    });
  } else {
    if (nodeMap.Query) {
      queryTypeName = 'Query';
    }
    if (nodeMap.Mutation) {
      mutationTypeName = 'Mutation';
    }
    if (nodeMap.Subscription) {
      subscriptionTypeName = 'Subscription';
    }
  }

  if (!queryTypeName) {
    errors.push(newGQLError(
      'Must provide schema definition with query type or a type named Query.',
      null,
      SEVERITY.error,
    ));
  }

  const innerTypeMap = {
    String: GQLString,
    Int: GQLInt,
    Float: GQLFloat,
    Boolean: GQLBoolean,
    ID: GQLID,
  };

  const types = typeDefs.map(def => typeDefNamed(def.name.value, def)).filter(Boolean);

  // Adding default types
  types.push(
    innerTypeMap.String,
    innerTypeMap.Int,
    innerTypeMap.Float,
    innerTypeMap.Boolean,
    innerTypeMap.ID,
  );

  // directives
  const directives = directiveDefs.map(getDirective);

  // If specified directives were not explicitly declared, add them.
  if (!directives.some(directive => directive.name === 'skip')) {
    directives.push(GQLSkipDirective);
  }

  if (!directives.some(directive => directive.name === 'include')) {
    directives.push(GQLIncludeDirective);
  }

  if (!directives.some(directive => directive.name === 'deprecated')) {
    directives.push(GQLDeprecatedDirective);
  }

  const schema = new GQLSchema({
    query: queryTypeName ? getObjectType(nodeMap[queryTypeName]) : null,
    mutation: mutationTypeName ? getObjectType(nodeMap[mutationTypeName]) : null,
    subscription: subscriptionTypeName ? getObjectType(nodeMap[subscriptionTypeName]) : null,
    types,
    directives,
    nodeMap,
  });

  return { schema, errors: [...errors, ...schema._errors] };

  function getDirective(directiveNode: DirectiveDefinitionNode): GQLDirective {
    return new GQLDirective(directiveNode, {
      name: directiveNode.name.value,
      description: getDescription(directiveNode),
      locations: directiveNode.locations.map(
        node => ((node.value: any): DirectiveLocationEnum),
      ),
      args: directiveNode.arguments && makeInputValues(directiveNode.arguments),
    });
  }

  function getObjectType(typeNode: TypeDefinitionNode): ?GQLObjectType {
    const type = typeDefNamed(typeNode.name.value, typeNode);

    if (!(type instanceof GraphQLObjectType)) {
      errors.push(newGQLError(
        'AST must provide object type.',
        [typeNode],
        SEVERITY.error,
      ));
    }

    return (type: any);
  }

  function typeDefNamed(typeName: string, node: ASTNode): ?GQLNamedType {
    if (innerTypeMap[typeName]) {
      if (innerTypeMap[typeName].dependents) {
        innerTypeMap[typeName].dependents.push(node);
      }
      return innerTypeMap[typeName];
    }

    if (!nodeMap[typeName]) {
      // NOT found
      errors.push(newGQLError(
        `Type "${typeName}" not found.`,
        [node],
        SEVERITY.error,
      ));
      return null;
    }

    const innerTypeDef = makeSchemaDef(nodeMap[typeName]);
    innerTypeDef.node = nodeMap[typeName]; // add location info
    innerTypeMap[typeName] = innerTypeDef;
    return innerTypeDef;
  }

  // function isProducedAlready(typeNode: TypeNode) {
  //   return Boolean(innerTypeMap[getNamedTypeNode(typeNode).name.value]);
  // }

  function produceType(typeNode: TypeNode, defaultValue: (name: string) => GraphQLType): GQLType {
    const namedTypeNode = getNamedTypeNode(typeNode);
    const typeName = namedTypeNode.name.value;
    const typeDef = typeDefNamed(typeName, namedTypeNode) || defaultValue(typeName);
    return buildWrappedType(typeDef, typeNode);
  }

  function produceInputType(typeNode: TypeNode): GraphQLInputType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.inputType);
    if (!isInputType((type: any))) {
      errors.push(newGQLError(
        'Expected Input type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function produceOutputType(typeNode: TypeNode): GraphQLOutputType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.outputType);
    if (!isOutputType((type: any))) {
      errors.push(newGQLError(
        'Expected Output type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function produceObjectType(typeNode: TypeNode): GQLObjectType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.objectType);
    if (!(type instanceof GraphQLObjectType)) {
      errors.push(newGQLError(
        'Expected Object type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function produceInterfaceType(typeNode: TypeNode): GQLInterfaceType {
    const type = produceType(typeNode, PLACEHOLDER_TYPES.interfaceType);
    if (!(type instanceof GraphQLInterfaceType)) {
      errors.push(newGQLError(
        'Expected Interface type.',
        [getNamedTypeNode(typeNode)],
        SEVERITY.error,
      ));
    }
    return (type: any);
  }

  function makeSchemaDef(def) {
    if (!def) {
      throw new Error('def must be defined');
    }
    switch (def.kind) {
      case OBJECT_TYPE_DEFINITION:
        return makeTypeDef(def);
      case INTERFACE_TYPE_DEFINITION:
        return makeInterfaceDef(def);
      case ENUM_TYPE_DEFINITION:
        return makeEnumDef(def);
      case UNION_TYPE_DEFINITION:
        return makeUnionDef(def);
      case SCALAR_TYPE_DEFINITION:
        return makeScalarDef(def);
      case INPUT_OBJECT_TYPE_DEFINITION:
        return makeInputObjectDef(def);
      default:
        throw new Error(`Type kind "${def.kind}" not supported.`);
    }
  }

  function makeTypeDef(def: ObjectTypeDefinitionNode) {
    const typeName = def.name.value;
    return new GQLObjectType(def, {
      name: typeName,
      description: getDescription(def),
      fields: () => makeFieldDefMap(def),
      interfaces: () => makeImplementedInterfaces(def),
    });
  }

  function makeFieldDefMap(
    def: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
  ) {
    return keyValMap(
      def.fields,
      field => field.name.value,
      field => ({
        type: produceOutputType(field.type),
        node: field,
        description: getDescription(field),
        args: makeInputValues(field.arguments),
        deprecationReason: getDeprecationReason(field.directives),
      }),
    );
  }

  function makeImplementedInterfaces(def: ObjectTypeDefinitionNode) {
    return def.interfaces && def.interfaces.map(iface => produceInterfaceType(iface));
  }

  function makeInputValues(values: Array<InputValueDefinitionNode>) {
    return keyValMap(
      values,
      value => value.name.value,
      (value) => {
        const type = produceInputType(value.type);
        return {
          type,
          node: value,
          description: getDescription(value),
          defaultValue: valueFromAST(value.defaultValue, type),
        };
      },
    );
  }

  function makeInterfaceDef(def: InterfaceTypeDefinitionNode) {
    const typeName = def.name.value;
    return new GQLInterfaceType(def, {
      name: typeName,
      description: getDescription(def),
      fields: () => makeFieldDefMap(def),
      resolveType: cannotExecuteSchema,
    });
  }

  function makeEnumDef(def: EnumTypeDefinitionNode) {
    const enumType = new GQLEnumType(def, {
      name: def.name.value,
      description: getDescription(def),
      values: keyValMap(
        def.values,
        enumValue => enumValue.name.value,
        enumValue => ({
          description: getDescription(enumValue),
          deprecationReason: getDeprecationReason(enumValue.directives),
          node: enumValue,
        }),
      ),
    });

    return enumType;
  }

  function makeUnionDef(def: UnionTypeDefinitionNode) {
    return new GQLUnionType(def, {
      name: def.name.value,
      description: getDescription(def),
      types: def.types.map(t => produceObjectType(t)),
      resolveType: cannotExecuteSchema,
    });
  }

  function makeScalarDef(def: ScalarTypeDefinitionNode) {
    return new GQLScalarType(def, {
      name: def.name.value,
      description: getDescription(def),
      serialize: () => null,
      // Note: validation calls the parse functions to determine if a
      // literal value is correct. Returning null would cause use of custom
      // scalars to always fail validation. Returning false causes them to
      // always pass validation.
      parseValue: () => false,
      parseLiteral: () => false,
    });
  }

  function makeInputObjectDef(def: InputObjectTypeDefinitionNode) {
    return new GQLInputObjectType(def, {
      name: def.name.value,
      description: getDescription(def),
      fields: () => makeInputValues(def.fields),
    });
  }
}

function getDeprecationReason(directives: ?Array<DirectiveNode>): ?string {
  const deprecatedAST = directives && find(
    directives,
    directive => directive.name.value === GQLDeprecatedDirective.name,
  );
  if (!deprecatedAST) {
    return null;
  }
  const { reason } = getArgumentValues(
    GQLDeprecatedDirective,
    deprecatedAST,
  );
  return (reason: any);
}

/**
 * Given an ast node, returns its string description based on a contiguous
 * block full-line of comments preceding it.
 */
export function getDescription(node: { loc?: Location }): ?string {
  const loc = node.loc;
  if (!loc) { return null; }
  const comments = [];
  let minSpaces;
  let token = loc.startToken.prev;
  while (
    token &&
    token.kind === TokenKind.COMMENT &&
    token.next && token.prev &&
    token.line + 1 === token.next.line &&
    token.line !== token.prev.line
  ) {
    const value = String(token.value);
    const spaces = leadingSpaces(value);
    if (minSpaces === undefined || spaces < minSpaces) {
      minSpaces = spaces;
    }
    comments.push(value);
    token = token.prev;
  }
  return comments
    .reverse()
    .map(comment => comment.slice(minSpaces))
    .join('\n');
}

// Count the number of spaces on the starting side of a string.
function leadingSpaces(str) {
  let i = 0;
  for (; i < str.length; i += 1) {
    if (str[i] !== ' ') {
      break;
    }
  }
  return i;
}

function cannotExecuteSchema() {
  throw new Error(
    'Generated Schema cannot use Interface or Union types for execution.',
  );
}
