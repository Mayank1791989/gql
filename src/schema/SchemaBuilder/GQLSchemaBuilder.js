/* @flow */
import path from 'path';
import fs from 'fs';

import { Source } from 'graphql/language/source';
import { parse } from 'graphql/language/parser';
import { GraphQLSchema } from 'graphql/type';

import { buildASTSchema } from './buildASTSchema';
import { buildASTSchema as buildASTGraphQLSchema } from 'graphql/utilities';

import { type GQLError, SEVERITY, toGQLError } from '../../shared/GQLError';

import { validate } from '../../schema';
import GQLConfig from '../../config/GQLConfig';

import watch from '../../shared/watch';

import { type ParsedFilesMap, type WatchFile } from '../../shared/types';

import { type GQLSchema } from '../../shared/GQLTypes';
import { type DocumentNode } from 'graphql/language/ast';

export default class GQLSchemaBuilder {
  _config: GQLConfig;
  _schema: GQLSchema;
  _ast: DocumentNode;
  _errors: Array<GQLError>;
  _parsedFilesMap: ParsedFilesMap = new Map();
  _isInitialized: boolean = false;

  constructor(options: {
    config: GQLConfig,
    onChange?: () => void,
    onInit?: Function,
    watch: boolean,
  }) {
    const { config, onChange, onInit } = options;

    this._config = config;

    // watch schema files and rebuild schema
    // console.log(config.getDir());
    const watchClient = watch({
      rootPath: config.getDir(),
      files: config.getSchema().files,
      name: 'gqlSchemaFiles',
      onChange: (files: Array<WatchFile>) => {
        this._updateFiles(files, config.getSchema());
        // console.log('init done');
        if (!this._isInitialized) {
          this._isInitialized = true;
          // console.log(this._isInitialized);
          if (onInit) {
            onInit();
          }
        }

        if (onChange) {
          onChange();
        }
        if (!options.watch) {
          watchClient.end();
        }
      },
    });
  }

  getSchema(): GQLSchema {
    return this._schema;
  }

  getGraphQLSchema(): GraphQLSchema {
    return buildASTGraphQLSchema(this._ast);
  }

  getSchemaErrors(): Array<GQLError> {
    return this._errors;
  }

  // private methods
  _updateFiles(files: Array<WatchFile>, config: any) {
    if (files.length === 0) {
      return;
    }

    // console.time('updating files');
    files.forEach(({ name, exists }) => {
      // console.log(name, exists);
      const absPath = path.join(this._config.getDir(), name);
      if (exists) {
        this._parsedFilesMap.set(absPath, this._parseFile(absPath));
      } else {
        this._parsedFilesMap.delete(absPath);
      }
    });
    // console.timeEnd('updating files');

    //  build merged ast
    // console.time('buildAST');
    const { ast, errors: parseErrors } = this._buildASTFromParsedFiles(
      this._parsedFilesMap,
    );
    // console.timeEnd('buildAST');

    // build GQLSchema from ast
    // console.time('buildASTSchema');
    const { schema, errors: buildErrors } = buildASTSchema(ast);
    // console.timeEnd('buildASTSchema');

    // validate
    // console.time('validate');
    // console.log(config.validate);
    const validationErrors = validate(schema, ast, config.validate);
    // console.timeEnd('validate');

    this._ast = ast;
    this._schema = schema;
    this._errors = [...parseErrors, ...buildErrors, ...validationErrors];
  }

  _parseFile = (absPath: string) => {
    const content = fs.readFileSync(absPath, 'utf8');
    const source = new Source(content, absPath);
    try {
      const ast = parse(source);
      return {
        ast,
        error: null,
      };
    } catch (err) {
      return {
        error: toGQLError(err, SEVERITY.error),
        ast: null,
      };
    }
  };

  _buildASTFromParsedFiles = (parsedFiles: ParsedFilesMap) => {
    const mergedDefinitions = [];
    const errors = [];
    for (const parsedFile of parsedFiles.values()) {
      // eslint-disable-line no-restricted-syntax
      const { ast, error } = parsedFile;
      if (error) {
        errors.push(error);
      } else {
        const { definitions } = ast;
        mergedDefinitions.push(...definitions);
      }
    }

    return {
      errors, // parsed errors
      ast: {
        kind: 'Document',
        definitions: mergedDefinitions,
      },
    };
  };
}
