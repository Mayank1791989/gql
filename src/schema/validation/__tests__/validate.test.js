/* @flow */
import { parse } from 'graphql/language/parser';
import { buildASTSchema } from '../../SchemaBuilder/buildASTSchema';
import { validate } from '../validate';

it('Report unused type definition', () => {
  const ast = parse(`
    type A {
      name: String
    }

    type X {
      value: Int!
    }

    type B {
      name: A
    }
  `);

  const { schema } = buildASTSchema(ast);
  const errors = validate(schema, ast);
  expect(errors).toMatchSnapshot();
});
