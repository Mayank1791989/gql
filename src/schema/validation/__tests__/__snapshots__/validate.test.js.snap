exports[`test Report unused type definition 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 5,
        "line": 6,
        "path": "GraphQL",
      },
    ],
    "message": "Unused type definition \'X\'",
    "severity": "warn",
  },
  Object {
    "locations": Array [
      Object {
        "column": 5,
        "line": 10,
        "path": "GraphQL",
      },
    ],
    "message": "Unused type definition \'B\'",
    "severity": "warn",
  },
]
`;
